### ISSUE TITLE
Make buildPhases (&c) composable

### ISSUE THREAD

**Issue by SomeoneSerge** [NixOS/nixpkgs#296544]
## Issue description

Setup hooks defining generic implementations for the common phases (`configurePhase`, `buildPhase`, `installPhase`, etc) weren't designed to be composed: `cmake`'s setup-hook only injects its implementation of the `configurePhase` if the "slot" hasn't been occupied yet by someone else (e.g. meson). Moreover, since pre- and post-hooks are usually executed by the phase implementations, one couldn't "compose" different implementations of the same phase even manually without running the same hooks twice.

For instance, it's currently hard to reuse the functionality of `cmakeBuildPhase` and `pypaBuildPhase` in the same derivation. Here's a mildly cursed example of a package generating a `setup.py` using `cmake` where the phases have been composed "manually": https://github.com/SomeoneSerge/dust3r.nix/blob/4fd7d61310f2289ba9843c9854aaaf59837eafbd/nix/magnum-bindings/package.nix#L41-L56.

The particular issue with `runHook`s could be addressed by somehow moving the `runHook`s to `generic/setup.sh`, but more importantly we need a discussion on how to better accommodate situations where parts of setup-hooks modeling common tools (cmake, meson, pypa-build, etc) need to be reused

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

**Comment by lolbinarycat** [comment]
possible solution: replace `buildPhase=pypaBuildPhase` with `postBuildHooks+=(pypaBuildPhase)`

would likely need other changes too.

**Comment by SomeoneSerge** [comment]
> possible solution: replace buildPhase=pypaBuildPhase with postBuildHooks+=(pypaBuildPhase)
> would likely need other changes too.

Yes it would, because as is this should result in an infinite recursion (pypaBuildPhase executes runHook postBuild, which runs pypaBuildPhase)

**Comment by lolbinarycat** [comment]
do we need a different primitive, or do we just need to remove the `runHook`?

**Comment by SomeoneSerge** [comment]
> do we need a different primitive, or do we just need to remove the runHook?

All of the other phases (including those potentially implemented out of tree) presumably use the same pattern

**Comment by ShamrockLee** [comment]
What about defining Bash functions containing the logic of a phase minus the pre/post hooks?
- #298838
